### PR TITLE
[WIP] add optional boolean args "prepend" to cumsum, cumprod

### LINKED ISSUE #76191: [Feature request] Exclusive prefix sum, `torch.cumsum(input, dim=0, exclusive=True)`
### 🚀 The feature, motivation and pitch

Currently `torch.cumsum` only supports inclusive prefix sum. However, the exclusive prefix sum is also needed in some cases.

Proposed to add a `exclusive` parameter to the `torch.cumsum` function
```py
input = torch.tensor([1, 1, 1])
torch.cumsum(input, dim=0, exclusive=True)   # [1, 2, 3]
torch.cumsum(input, dim=0, exclusive=False)  # [0, 1, 2]
```

Related issue from PyTorch Forums:
- [PyTorch equivalent of exclusive cumsum? - PyTorch Forums](https://discuss.pytorch.org/t/pytorch-equivalent-of-exclusive-cumsum/107259)

CUDA implementation from CUB:
- [CUB: cub::DeviceScan Struct Reference](https://nvlabs.github.io/cub/structcub_1_1_device_scan.html#a02b2d2e98f89f80813460f6a6ea1692b)

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128469
* __->__ #127675
* #128299

It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Differential Revision: [D58329265](https://our.internmc.facebook.com/intern/diff/D58329265)

### CODE REVIEW

**Comment by davidberard98** [comment]
@BowenBao @thiagocrepaldi could you recommend how I should update ONNX to handle this additional arg?

**Comment by BowenBao** [comment]
cc @xadupre @titaiwangms for ONNX asks.

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -787,12 +787,14 @@ def _reduce_with_dtype(onnx_op: str, name: str, allow_multi_dim_support: bool =
 
 
 @_onnx_symbolic("aten::cumsum")
-@symbolic_helper.parse_args("v", "i", "none")
+@symbolic_helper.parse_args("v", "i", "none", "b")
 @_beartype.beartype
-def cumsum(g: jit_utils.GraphContext, input, dim, dtype):
+def cumsum(g: jit_utils.GraphContext, input, dim, dtype, full):
```
Comment (titaiwangms) [inline torch/onnx/symbolic_opset9.py:792]: Is it because this one does not have default value as the other one, so it's caught by BC_linter?

**Comment by titaiwangms** [comment]
For the test failure in `onnx/test_fx_to_onnx_with_onnxruntime.py` with the error message:

```
torch.onnx._internal.diagnostics.infra.context.RuntimeErrorWithDiagnostic: Cannot find any perfect/nearest match of symbolic function for aten::cumsum.default,which should be registered under aten.cumsum.default.
```

Please xfail or skip them, as the symbolic function for onnx.dynamo_export is not yet in PyTorch.

cc @xadupre @justinchuby @shubhambhokare1

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58329265).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58329265).
